File: 11-Binary-Search/cpp/08-aggressive-cows.cpp
class Solution {
public:
    int aggressiveCows(vector<int>& nums, int k) {
        sort(nums.begin(), nums.end());
        int n = nums.size();

        int low = 1;
        int high = nums[n - 1] - nums[0];
        int ans = 0;

        while (low <= high) {
            int mid = low + (high - low) / 2;

            if (canPlaceCows(nums, k, mid)) {
                ans = mid;
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }

        return ans;
    }

private:
    bool canPlaceCows(const vector<int>& nums, int k, int minDist) {
        int cowsPlaced = 1;
        int lastPosition = nums[0];

        for (int i = 1; i < (int)nums.size(); i++) {
            if (nums[i] - lastPosition >= minDist) {
                cowsPlaced++;
                lastPosition = nums[i];
            }

            if (cowsPlaced >= k) {
                return true;
            }
        }

        return false;
    }
};

File: 11-Binary-Search/python/08_aggressive_cows.py
class Solution:
    def aggressiveCows(self, nums: List[int], k: int) -> int:
        nums.sort()
        n = len(nums)

        low = 1
        high = nums[n - 1] - nums[0]
        ans = 0

        def can_place_cows(min_dist):
            cows_placed = 1
            last_position = nums[0]
            for i in range(1, len(nums)):
                if nums[i] - last_position >= min_dist:
                    cows_placed += 1
                    last_position = nums[i]
                if cows_placed >= k:
                    return True
            return False

        while low <= high:
            mid = (low + high) // 2
            if can_place_cows(mid):
                ans = mid
                low = mid + 1
            else:
                high = mid - 1

        return ans
